File: Python/DFS/TreeDiameter.py
# @author
# Aakash Verma

# Output:
# 8

from collections import *

class Node:
    def __init__(self, data):
        self.data = data
        self.right = self.left = None

class TreeDiameter:
    
    diameter = 0

    def __init__(self): 
        self.root = None

    def findHeight(self, root):
        if root is None: return 0

        leftHeight = self.findHeight(root.left)
        rightHeight = self.findHeight(root.right)

        currentDiameter = leftHeight + rightHeight + 1

        self.diameter = max(currentDiameter, self.diameter)

        return max(leftHeight, rightHeight) + 1

    def findDiameter(self, root):
        self.findHeight(root)
        return self.diameter
        
        
if __name__ == '__main__':
    tree = TreeDiameter()
    tree.root = Node(1)
    tree.root.left = Node(2)
    tree.root.right = Node(3)
    tree.root.right.left = Node(4)
    tree.root.right.right = Node(5)
    tree.root.right.left.left = Node(6)
    tree.root.right.right.right = Node(7)
    tree.root.right.right.right.left = Node(12)
    tree.root.right.right.right.right = Node(13)
    tree.root.right.left.left.left = Node(8)
    tree.root.right.left.left.right = Node(10)
    tree.root.right.left.left.right.left = Node(11)
    ans = tree.findDiameter(tree.root)
    print(ans)

File: Java/DFS/TreeDiameter.java
/*
*   @author
*   Aakash Verma
*
*   Output:
*   8
*
*/

import java.util.*;

class Node {
    int data;
    Node left, right;
    public Node(int key) {
        data = key;
        left = right = null;
    }
}


class TreeDiameter {

    static Node root;
    TreeDiameter() {
        root = null;
    }

    static int diameter = 0;

    private static int findHeight(Node root) {
        if(root == null) return 0;

        int leftHeight = findHeight(root.left);
        int rightHeight = findHeight(root.right);
        
        int currentDiameter = leftHeight + rightHeight + 1;

        diameter = Math.max(currentDiameter, diameter);
        return Math.max(leftHeight, rightHeight) + 1;
    }

    public static int findDiameter(Node root) {
        findHeight(root);
        return diameter;
    }

    public static void main (String[] args) {
        root = new Node(1);
        root.left = new Node(2);
        root.right = new Node(3);
        root.right.left = new Node(4);
        root.right.right = new Node(5);
        root.right.left.left = new Node(6);
        root.right.right.right = new Node(7);
        root.right.right.right.left = new Node(12);
        root.right.right.right.right = new Node(13);
        root.right.left.left.left = new Node(8);
        root.right.left.left.right = new Node(10);
        root.right.left.left.right.left = new Node(11);
        int ans = findDiameter(root);
        System.out.println(ans);
    }
}